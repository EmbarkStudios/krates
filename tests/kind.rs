mod util;

use util::build;

#[test]
fn all_the_things() {
    let grafs = build("all-features.json", krates::Builder::new()).unwrap();

    assert!(grafs.filtered.is_empty());
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn filters_dev() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        // This shouldn't actually affect anything, as dev dependencies
        // for non-workspace crates are already not resolved
        assert!(grafs.filtered.is_empty());
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);

        // This will be equivalent to to filtering workspace dev crates
        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_build() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_normal() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Normal, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn filters_build_and_dev() {
    // Just non-workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Just workspace crates
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::Workspace);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::Workspace);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }

    // Both
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();
        insta::assert_snapshot!(grafs.dotgraph());
    }
}

#[test]
fn only_b() {
    let mut kb = krates::Builder::new();
    kb.include_workspace_crates(["/home/jake/code/krates/tests/ws/b/Cargo.toml"]);

    let grafs = build("all-features.json", kb).unwrap();
    insta::assert_snapshot!(grafs.dotgraph());
}

#[test]
fn filters_after_build() {
    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        let filtered = grafs.actual.krates_filtered(krates::DepKind::Dev);

        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Dev, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();

        let expected = format!("{:#?}", grafs.actual.krates().collect::<Vec<_>>());
        let actual = format!("{filtered:#?}");

        similar_asserts::assert_eq!(expected, actual);
    }

    {
        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::NonWorkspace);

        let grafs = build("all-features.json", kb).unwrap();

        let filtered = grafs.actual.krates_filtered(krates::DepKind::Build);

        let mut kb = krates::Builder::new();
        kb.ignore_kind(krates::DepKind::Build, krates::Scope::All);

        let grafs = build("all-features.json", kb).unwrap();

        let expected = format!("{:#?}", grafs.actual.krates().collect::<Vec<_>>());
        let actual = format!("{filtered:#?}");

        similar_asserts::assert_eq!(expected, actual);
    }
}
