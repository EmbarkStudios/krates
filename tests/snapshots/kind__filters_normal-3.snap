---
source: tests/kind.rs
expression: grafs.dotgraph()
---
digraph {
    0 [ label = "crate a 0.1.0 path+file:///krates/tests/ws/a" ]
    1 [ label = "crate b 0.1.0 path+file:///krates/tests/ws/b" ]
    2 [ label = "crate c 0.1.0 path+file:///krates/tests/ws/c" ]
    3 [ label = "crate cc 1.0.73 git+https://github.com/alexcrichton/cc-rs#53fb72c87e5769a299f1886ead831901b9c775d6" ]
    4 [ label = "crate cc 1.0.73" ]
    5 [ label = "crate difference 2.0.0" ]
    6 [ label = "crate ring 0.16.20" ]
    7 [ label = "feature default" ]
    8 [ label = "feature default" ]
    9 [ label = "feature default" ]
    10 [ label = "feature leftpad" ]
    11 [ label = "feature leftier-strings" ]
    12 [ label = "feature lazy_static" ]
    13 [ label = "feature once_cell" ]
    14 [ label = "feature dev_urandom_fallback" ]
    15 [ label = "feature alloc" ]
    0 -> 7 [ label = "(dev)" ]
    0 -> 7 [ label = "(build) 'cfg(target_os = \"linux\")'" ]
    1 -> 8 [ label = "(dev) 'cfg(target_arch = \"x86_64\")'" ]
    2 -> 3 [ label = "(build)" ]
    2 -> 9 [ label = "(dev)" ]
    6 -> 4 [ label = "(build)" ]
    10 -> 2 [ label = "" ]
    11 -> 2 [ label = "" ]
    11 -> 10 [ label = "" ]
    12 -> 2 [ label = "" ]
    7 -> 2 [ label = "" ]
    7 -> 10 [ label = "" ]
    9 -> 5 [ label = "" ]
    13 -> 6 [ label = "" ]
    14 -> 6 [ label = "" ]
    14 -> 13 [ label = "" ]
    8 -> 6 [ label = "" ]
    8 -> 15 [ label = "" ]
    8 -> 14 [ label = "" ]
    15 -> 6 [ label = "" ]
}

